src/protocol/ipa/mod.rs

```diff
@@ -771,132 +771,195 @@ pub mod tests {
-                })
+        let result: Vec<GenericReportTestInput<Fp32BitPrime, MatchKey, BreakdownKey>> = world
+            .semi_honest(records, |ctx, input_rows| async move {
+                ipa::<Fp32BitPrime, MatchKey, BreakdownKey>(
+                    ctx,
+                    &input_rows,
+                    per_user_cap,
+                    max_breakdown_key as u128,
+                    NUM_MULTI_BITS,
+                )
                 .await
-                .reconstruct();
+                .unwrap()
+            })
+            .await
+            .reconstruct();
 
-            assert_eq!(MAX_BREAKDOWN_KEY, result.len());
-            println!(
-                "actual results: {:#?}",
-                result
-                    .iter()
-                    .map(|x| x.trigger_value.as_u128())
-                    .collect::<Vec<_>>(),
+        assert_eq!(max_breakdown_key, result.len());
+        println!(
+            "actual results: {:#?}",
+            result
+                .iter()
+                .map(|x| x.trigger_value.as_u128())
+                .collect::<Vec<_>>(),
+        );
+        for (i, expected) in expected_results.iter().enumerate() {
+            assert_eq!(
+                [i as u128, u128::from(*expected)],
+                [
+                    result[i].breakdown_key.as_u128(),
+                    result[i].trigger_value.as_u128()
+                ]
             );
-            for (i, expected) in expected_results.iter().enumerate() {
-                assert_eq!(
-                    [i as u128, u128::from(*expected)],
-                    [
-                        result[i].breakdown_key.as_u128(),
-                        result[i].trigger_value.as_u128()
-                    ]
+        }
+    }
+
+    #[tokio::test]
+    #[allow(clippy::missing_panics_doc)]
+    pub async fn random_ipa_check() {
+        const MAX_BREAKDOWN_KEY: usize = 16;
+        const MAX_TRIGGER_VALUE: u32 = 5;
+        const NUM_USERS: usize = 10;
+        const MAX_RECORDS_PER_USER: usize = 8;
+        const NUM_MULTI_BITS: u32 = 3;
+
+        let random_seed = thread_rng().gen();
+        println!("Using random seed: {random_seed}");
+        let mut rng = StdRng::seed_from_u64(random_seed);
+
+        let mut random_user_records = Vec::with_capacity(NUM_USERS);
+        for _ in 0..NUM_USERS {
+            let records_for_user = generate_random_user_records_in_reverse_chronological_order(
+                &mut rng,
+                MAX_RECORDS_PER_USER,
+                MAX_BREAKDOWN_KEY,
+                MAX_TRIGGER_VALUE,
+            );
+            random_user_records.push(records_for_user);
+        }
+        let mut raw_data = random_user_records.concat();
+
+        // Sort the records in chronological order
+        // This is part of the IPA spec. Callers should do this before sending a batch of records in for processing.
+        raw_data.sort_unstable_by(|a, b| a.timestamp.cmp(&b.timestamp));
```

Comment:
One thing that we might loosen as a constraint here is that maybe the caller only needs to sort those events that *might* be from the same person.  That means that if they know records are from different people, they can ignore the relative sort order. That might make certain types of sorting more efficient.

I was realized that this test would work fine without this step.  Of course, then our sort doesn't really do anything of use, so of course we should do the full sort in our test case...

Reply:
That is an interesting observation. I'm not sure how we can utilize it - but I really hope we can find a way!

If you're running source-fanout query, and you are showing ads to signed-in users, you can easily segment your source-events into multiple groups of non-overlapping users. Let's say 100 groups. If you had a load of trigger-events you wanted to try to join up with them, I suppose you could just add them to ALL of the groups, and then sort all 100 groups separately (and do attribution / capping / aggregation separately as well). Any given trigger event could only affect one of those groups, so there should be no double counting (might need some kind of proof that it was segmented right to be sure tho...)

This would *definitely* be faster if trigger events are relatively rare compared to source events. It would *definitely* be a lot of extra work in the case that source and trigger events come in equal volumes.